### ISSUE TITLE
macOS: call to undeclared function '_NSGetExecutablePath'

### ISSUE THREAD

**Issue by cbenhagen** [Nuitka/Nuitka#2287]
Nuitka 1.6.3 and some versions before fail to compile on macOS due to a missing include in `HelpersFilesystemPaths.c`:

```
In file included from static_src/CompiledFunctionType.c:2864:
In file included from /Users/ben/Library/Caches/pypoetry/virtualenvs/myapp-NW6c3THU-py3.11/lib/python3.11/site-packages/nuitka/build/static_src/CompiledCodeHelpers.c:70:
/Users/ben/Library/Caches/pypoetry/virtualenvs/myapp-NW6c3THU-py3.11/lib/python3.11/site-packages/nuitka/build/static_src/HelpersFilesystemPaths.c:40:15: error: call to undeclared function '_NSGetExecutablePath'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    int res = _NSGetExecutablePath(binary_filename, &bufsize);
              ^
1 error generated.
```

I am re-creating this issue for others who google for it as my original one seems to have been accidentaly deleted by the maintainers. The issue has already been fixed in 15b5d37d1010b36e638d3facbbaec0143daaff2c so I will close immediately after opening.

Workarounds are:
1. Install factory version
`pip install -U https://github.com/Nuitka/Nuitka/archive/factory.zip`

2. Add the following lines to the top of `HelpersFilesystemPaths.c`:
```C
#if defined(__APPLE__)
#include <mach-o/dyld.h>
#endif
```

**Comment by cbenhagen** [comment]
Fixed in 15b5d37d1010b36e638d3facbbaec0143daaff2c

**Comment by kayhayen** [comment]
I never reproduced this, and you never produced anything that would have allowed this, so I deleted it as noise, esp. as it referred to a factory state, which is explicitely asked to not report to the general tracker, but via Discord or not at all.

The error you describe was never observed, and you are wasting my time by pretending to report against 1.6.3 and then quoting an error line that doesn't exist in its code. So you wasted my time now. I am telling you, for the last time. Because also this applies:

In the template, I have made it clear, that not filling out the issue template is not acceptable. I am closing this kind of report as invalid. The main reason is that I do not want to be left guessing. For you it's easy to erase the issue template and expect me to ask you questions only as necessary, but I cannot do that. I would much rather wait for you to create a new report, fill out the template, or wait for somebody with the necessary skill to do it.

In short: No issue template, no fun. Therefore I am likely just deleting the issue, as it serves as a bad example and detraction for searches.  And even worse, it might keep good reporters from reporting as is this "already known."

**Comment by cbenhagen** [comment]
Hi @kayhayen

I was under the impression that the original issue included every information requested by the template when I removed the template itself. I am very sorry if I missed something critical but I can assure you that I have put a lot of effort into the issue to make sure it is useful and helps making nuitka even better for everyone. 

Before creating the issue I also made sure that it also occurs on a very simple project with a single hello world file, is present on different machines and that it was not already fixed in the latest development release as suggested in the template and is just common sense. I am sorry for not mentioning that it also exists in the stable release. That would have helped I guess. But as most projects prefer to only get issues that are present in the latest main branch I guess I was just used to leave this info out. This was corrected in the second issue but as I saw it was already fixed I then left out critical information like the macOS version and other details. Rest assured, that at no point I intended to waste anyones time. Quite the opposite is true in fact.

Just deleting it was perceived as very rude at first, hence why I assumed it must have happened by accident. But I also understand that having to handle too many low effort issues is very tiring and that it might have just cought you on the wrong foot. So no offence taken. 

While further reasearching this, I found that on one of my machines the issue is not present and after digging through the differences in between those machines I think I have found the one detail that is needed to reproduce it:
**LLVM version 16.0.5 installed through homebrew.**

```
/opt/homebrew/opt/llvm/bin/clang --version
Homebrew clang version 16.0.5
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /opt/homebrew/opt/llvm/bin
```

As opposed to the clang version 14.0.3 which nuitka would have used otherwise:
```
clang --version
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

After uninstalling it, the hello world compiled as expected. 

For the sake of completeness I might add that the machine I am currently testing on runs macOS 13.4 and Python 3.11.4 installed using the installer from python.org.

Nuitka version info:
```
python -m nuitka --version
1.6.3
Commercial: None
Python: 3.11.4 (v3.11.4:d2340ef257, Jun  6 2023, 19:15:51) [Clang 13.0.0 (clang-1300.0.29.30)]
Flavor: CPython Official
Executable: /Users/ben/Library/Caches/pypoetry/virtualenvs/bla-F7d9cPuG-py3.11/bin/python
OS: Darwin
Arch: arm64
Version C compiler: /opt/homebrew/opt/llvm/bin/clang (clang).
```

Yes that piece of info might have given a clue to what the issue might be. Even as it is missing the clang version. Sorry again for missing this in the first place. As this is not happening on the latest factory release anymore, I guess the issue can remain closed.

**Comment by kayhayen** [comment]
If I had seen this one, it would have been very relevant, I don't think you had this before.

Sorry if I gave you a hard time. Understand where I am coming from, people with skills, no skills, respect and no respect. There is a reason I am asking for the information in the issue template:

> Version C compiler: /opt/homebrew/opt/llvm/bin/clang (clang).

That is just plain different from what Apple does. Homebrew in general is supposed to work somewhat. It limits the platforms that are going to work, over what CPython and Anaconda do. And I am not sure it builds things in a ways that match what PyPI does well enough anyway. When using poetry to install things, it's probably not going to matter much. I wonder if "Flavor: CPython Official" is a misdetection there or if this is really a Homebrew Python, but maybe not recognized as such.

Question I have to myself, if Nuitka should accept the homebrew clang at all, or skip it. On macOS, the one from Xcode is probably the one to use in all cases. I don't see why we would want to use it. And in your case, it may even be non-intentional. Maybe it has less capabilties, or ties things closer to Homebrew than necessary, or maybe it's the same. It either way would appear to be unnecessary.

I think about the deleting, I explained myself. Bad precedent is seen by others. And people starting posting they have "the same" issue when in fact neither do they, nor do they know what issue it is, and that happens sometimes years later, where good reports with issue templates would be actually needed. And where they don't, they tend to think, it's not needed, but that's often not the case.

Clever people deciding what I need, and not so clever people being lazy are indistinguable from my point of view. There is always a new chance for those that care to improve. I write a template comment before I delete, which explains that.

For me, time is limited. I do not really want to guess anything. And when I find repeatedly that information is missing, like in this instance, I make it output the compiler chosen, because it matters. Clever people know it doesn't matter. Cleverness is wasting my time. In my mind, really clever people just paste the information I am asking, even if it shouldn't matter, because they need to help me to do it. Need to make a posting like this.

**Comment by kayhayen** [comment]
I think, this issue should be about avoiding to use Homebrew clang, and enforce Xcode clang, at least by default. I need to figure out how to best to that. It lives in a couple of places I believe.
